File: typescript/src/runtime/ancestry.ts
import { diag } from '@opentelemetry/api';
import type { SpanContext } from '@opentelemetry/api';
import type { ReadableSpan } from '@opentelemetry/sdk-trace-base';

export const MAX_PENDING_SPANS = 1024;
export const MAX_PENDING_BYTES = 16 * 1024 * 1024;
export const spanKey = (ctx: SpanContext): string =>
  `${ctx.traceId}:${ctx.spanId}`;

/** Conservative payload accounting, not an RSS bound; shared values count again. */
function size(value: unknown): number {
  if (typeof value === 'string')
    return 64 + 2 * Buffer.byteLength(value, 'utf8');
  if (Array.isArray(value))
    return 64 + value.reduce((n, v) => n + 16 + size(v), 0);
  if (value && typeof value === 'object')
    return (
      128 +
      Object.entries(value).reduce((n, [k, v]) => n + 64 + size(k) + size(v), 0)
    );
  return 32;
}
function estimatedSize(span: ReadableSpan): number {
  return (
    2048 +
    size(span.name) +
    size(span.attributes) +
    size(span.events) +
    size(span.links) +
    size(span.resource.attributes) +
    size(span.instrumentationScope) +
    size(span.status)
  );
}
interface Node<R> {
  route: R;
  parent?: string | undefined;
  children: number;
  ended: boolean;
  selected: boolean;
  payload?: ReadableSpan | undefined;
  size: number;
}

/** No waits or timers; active descendants protect ancestry regardless of age. */
export class Ancestry<R extends { active: number }> {
  readonly nodes = new Map<string, Node<R>>();
  readonly pending = new Map<string, Node<R>>();
  pendingBytes = 0;
  constructor(
    private readonly publish: (route: R, span: ReadableSpan) => void,
  ) {}
  start(
    key: string,
    parent: string | undefined,
    route: R,
    selected: boolean,
  ): void {
    const ancestor = parent ? this.nodes.get(parent) : undefined;
    if (!ancestor || ancestor.route !== route) parent = undefined;
    else ancestor.children++;
    this.nodes.set(key, {
      route,
      parent,
      children: 0,
      ended: false,
      selected: false,
      size: 0,
    });
    if (selected) this.select(key);
  }
  private releasePayload(key: string, node: Node<R>): void {
    this.pending.delete(key);
    this.pendingBytes -= node.size;
    node.payload = undefined;
    node.size = 0;
  }
  private select(key: string | undefined): void {
    while (key) {
      const node = this.nodes.get(key);
      if (!node || node.selected) break;
      node.selected = true;
      if (node.payload) {
        const payload = node.payload;
        this.releasePayload(key, node);
        this.publish(node.route, payload);
      }
      key = node.parent;
    }
  }
  end(key: string, span: ReadableSpan, selected: boolean): void {
    const node = this.nodes.get(key);
    if (!node) return;
    node.ended = true;
    if (selected) this.select(key);
    if (node.selected) this.publish(node.route, span);
    else if (node.children) {
      node.payload = span;
      node.size = estimatedSize(span);
      this.pending.set(key, node);
      this.pendingBytes += node.size;
    }
    this.clean(key);
    if (
      this.pending.size > MAX_PENDING_SPANS ||
      this.pendingBytes > MAX_PENDING_BYTES
    )
      diag.debug('Ancestry buffer limit reached; exporting pending ancestors');
    while (
      this.pending.size > MAX_PENDING_SPANS ||
      this.pendingBytes > MAX_PENDING_BYTES
    )
      this.select(this.pending.keys().next().value!);
  }
  private clean(key: string | undefined): void {
    while (key) {
      const node = this.nodes.get(key);
      if (!node || !node.ended || node.children) break;
      this.releasePayload(key, node);
      this.nodes.delete(key);
      node.route.active--;
      key = node.parent;
      const parent = key ? this.nodes.get(key) : undefined;
      if (parent) parent.children--;
    }
  }
  flushPending(): void {
    while (this.pending.size) this.select(this.pending.keys().next().value!);
  }
  clear(): void {
    for (const node of this.nodes.values()) node.route.active--;
    this.nodes.clear();
    this.pending.clear();
    this.pendingBytes = 0;
  }
}

File: python/src/confident_trace/_core/ancestry.py
"""Destination-local ancestry selection over public OTel span callbacks."""

import logging
import sys
import time
from collections import OrderedDict
from dataclasses import dataclass

from .._attributes import SCOPE_NAME

log = logging.getLogger(SCOPE_NAME)
MAX_PENDING_SPANS = 1024
MAX_PENDING_BYTES = 16 * 1024 * 1024


def span_key(context):
    return context.trace_id, context.span_id


def estimated_size(span):
    """Conservative payload estimate, not a process RSS limit.

    Include container overhead and double worst-case UTF-8 string storage. Count shared
    values repeatedly, including resource/scope data, rather than undercounting.
    """

    def size(value):
        if isinstance(value, str):
            return sys.getsizeof(value) + 8 * len(value)
        if hasattr(value, "items"):
            return sys.getsizeof(value) + sum(
                size(k) + size(v) + 64 for k, v in value.items()
            )
        if isinstance(value, (tuple, list)):
            return sys.getsizeof(value) + sum(size(v) + 16 for v in value)
        return sys.getsizeof(value)

    result = 2048 + size(span.name) + size(span.attributes)
    result += size(getattr(span.resource, "attributes", {}))
    scope = span.instrumentation_scope
    if scope:
        result += size(scope.name) + size(scope.version) + size(scope.attributes)
    for event in span.events:
        result += 256 + size(event.name) + size(event.attributes)
    for link in span.links:
        result += 256 + size(link.attributes)
    result += size(span.status.description)
    return result


@dataclass
class Node:
    route: object
    parent: object = None
    children: int = 0
    ended: bool = False
    selected: bool = False
    payload: object = None
    size: int = 0


class Ancestry:
    """Caller serializes access; only ended unresolved payloads are buffered."""

    def __init__(self, publish):
        self.nodes = {}
        self.pending = OrderedDict()
        self.pending_bytes = 0
        self.publish = publish

    def start(self, key, parent, route, selected):
        ancestor = self.nodes.get(parent)
        if ancestor is None or ancestor.route is not route:
            parent = None
        else:
            ancestor.children += 1
        self.nodes[key] = Node(route, parent)
        if selected:
            self.select(key)

    def select(self, key):
        while key in self.nodes:
            node = self.nodes[key]
            if node.selected:
                break
            node.selected = True
            if node.payload is not None:
                payload = node.payload
                self._release_payload(key, node)
                self.publish(node.route, payload)
            key = node.parent

    def _release_payload(self, key, node):
        self.pending.pop(key, None)
        self.pending_bytes -= node.size
        node.payload, node.size = None, 0

    def end(self, key, span, selected):
        node = self.nodes.get(key)
        if node is None:
            return
        node.ended = True
        if selected:
            self.select(key)
        if node.selected:
            self.publish(node.route, span)
        elif node.children:
            node.payload, node.size = span, estimated_size(span)
            self.pending[key] = None
            self.pending_bytes += node.size
        self._clean(key)
        if (
            len(self.pending) > MAX_PENDING_SPANS
            or self.pending_bytes > MAX_PENDING_BYTES
        ):
            log.debug("Ancestry buffer limit reached; exporting pending ancestors")
        while (
            len(self.pending) > MAX_PENDING_SPANS
            or self.pending_bytes > MAX_PENDING_BYTES
        ):
            self.select(next(iter(self.pending)))

    def _clean(self, key):
        while key in self.nodes:
            node = self.nodes[key]
            if not node.ended or node.children:
                break
            self._release_payload(key, node)
            del self.nodes[key]
            node.route.active -= 1
            key = node.parent
            if key in self.nodes:
                self.nodes[key].children -= 1

    def flush_pending(self, deadline=None):
        while self.pending:
            if deadline is not None and time.monotonic() >= deadline:
                return False
            self.select(next(iter(self.pending)))
        return True

    def clear(self):
        for node in self.nodes.values():
            node.route.active -= 1
        self.nodes.clear()
        self.pending.clear()
        self.pending_bytes = 0
